azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -86,12 +98,115 @@ public ProjectManager(ProjectLoader loader, Props props) {
     new XmlValidatorManager(prop);
     loadAllProjects();
     loadProjectWhiteList();
+
+    CleanerThread cleanerThread = new CleanerThread();
+    cleanerThread.start();
+  }
+
+  /**
+   * Cleaner thread to clean projects. Runs every day.
+  **/
+  private class CleanerThread extends Thread {
+
+    // runs every day
+    private static final long CLEANER_THREAD_WAIT_INTERVAL_MS = 24 * 60 * 60 * 1000;
+    private static final String DELETER = "azkaban";
+
+    private boolean shutdown = false;
+    private long lastProjectFileCleanTime = -1;
+    private long lastProjectCleanTime = -1;
+
+    public CleanerThread() {
+      this.setName("ProjectManager-Cleaner-Thread");
+      logger.info("Starting cleaner thread: " + this.getName());
+    }
+
+    @SuppressWarnings("unused")
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
+    }
+
+    public void run() {
+      while (!shutdown) {
+        synchronized (this) {
+          try {
+            long currentTime = System.currentTimeMillis();
+
+            if (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectFileCleanTime) {
+              cleanProjectFilesAndRemoveProject();
+              lastProjectFileCleanTime = currentTime;
+            }
+
+            if (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectCleanTime) {
+              purgeRemovedProjects();
+              lastProjectCleanTime = currentTime;
+            }
+
+            logger.info("Cleaner Thread completed. Waiting!!!!!");
+            wait(CLEANER_THREAD_WAIT_INTERVAL_MS);
+          } catch (ProjectManagerException e) {
+            logger.error(this.getName() + "failed due to: " + e.getMessage());
+          } catch (InterruptedException e) {
+            logger.info("Interrupted. Probably to shut down.");
+          }
+        }
+      }
+    }
+
+    private void cleanProjectFilesAndRemoveProject() throws ProjectManagerException {
+      User deleter = new User(DELETER);
+      long currentTime = System.currentTimeMillis();
+
+      Set<Integer> scheduledProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.SCHEDULE);
+      logger.info("Scheduled Projects count: " + scheduledProjectIds.size());
+
+      List<Project> projects = projectsById.values()
+              .stream()
+              .filter(project -> !scheduledProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > projectFilesRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Project files to clean: " + projects.size());
+
+      for(Project project : projects) {
+        logger.debug("Cleaning project files for project: " + project.getId());
```

Comment:
log info perhaps as we are cleaning up ?

Reply:
We thought this may fill up the logs but changed it to INFO for now since this will be initially a large set then it will slowly decrease